libparsec/crates/client/tests/unit/certif/validate_block.rs

```diff
@@ -0,0 +1,889 @@
+            authenticated_cmds::latest::realm_get_keys_bundle::Rep::AccessNotAvailableForAuthor
+        },
+    );
+    let err = ops
+        .validate_block(
+            data.wksp1_last_realm_certificate_timestamp,
+            data.wksp1_id,
+            data.wksp1_last_key_index,
+            &data.wksp1_bar_txt_manifest,
+            &data.wksp1_bar_txt_block_access,
+            &data.wksp1_bar_txt_block_encrypted,
+        )
+        .await
+        .unwrap_err();
+
+    p_assert_matches!(err, CertifValidateBlockError::Stopped);
+}
+
+#[parsec_test(testbed = "minimal_client_ready")]
+async fn key_bundle_bad_key_index(env: &TestbedEnv) {
+    let data = BaseData::extract(env);
+    let alice = env.local_device("alice@dev1");
+    let ops = certificates_ops_factory(env, &alice).await;
+
+    test_register_send_hook(
+        &env.discriminant_dir,
+        move |_req: authenticated_cmds::latest::realm_get_keys_bundle::Req| {
+            authenticated_cmds::latest::realm_get_keys_bundle::Rep::BadKeyIndex
+        },
+    );
+    let err = ops
+        .validate_block(
+            data.wksp1_last_realm_certificate_timestamp,
+            data.wksp1_id,
+            data.wksp1_last_key_index,
+            &data.wksp1_bar_txt_manifest,
+            &data.wksp1_bar_txt_block_access,
+            &data.wksp1_bar_txt_block_encrypted,
+        )
+        .await
+        .unwrap_err();
+
+    p_assert_matches!(err, CertifValidateBlockError::Stopped);
+}
+
+// InvalidBlockAccess(#[from] Box<InvalidBlockAccessError>),
+// InvalidCertificate(#[from] Box<InvalidCertificateError>),
+// InvalidKeysBundle(#[from] Box<InvalidKeysBundleError>),
+// Internal(#[from] anyhow::Error),
+
+#[parsec_test(testbed = "minimal_client_ready")]
+async fn bad_digest(env: &TestbedEnv) {
+    let mut data = BaseData::extract(env);
+    let alice = env.local_device("alice@dev1");
+    let ops = certificates_ops_factory(env, &alice).await;
+
+    test_register_send_hook(
+        &env.discriminant_dir,
+        test_send_hook_realm_get_keys_bundle!(env, alice.user_id, data.wksp1_id),
+    );
+
+    data.wksp1_bar_txt_block_access.digest =
+        hex!("7d486915b914332bb5730fd772223e8b276919e51edca2de0f82c5fc1bce7eb6").into();
+    let err = ops
+        .validate_block(
+            data.wksp1_last_realm_certificate_timestamp,
+            data.wksp1_id,
+            data.wksp1_last_key_index,
+            &data.wksp1_bar_txt_manifest,
+            &data.wksp1_bar_txt_block_access,
+            &data.wksp1_bar_txt_block_encrypted,
+        )
+        .await
+        .unwrap_err();
+
+    p_assert_matches!(
+        err,
+        CertifValidateBlockError::InvalidBlockAccess(err)
+        if err.to_string() == "Block access `f0000000-0000-0000-0000-000000000004` from manifest `f0000000-0000-0000-0000-000000000005` version 1 (in realm `f0000000-0000-0000-0000-000000000002`, create by `de10a11c-ec00-1000-0000-000000000000` on 2000-01-07T00:00:00Z): hash digest mismatches !"
```

Comment:
Oops, need to change error string here too? ^^
```suggestion
        if err.to_string() == "Block access `f0000000-0000-0000-0000-000000000004` from manifest `f0000000-0000-0000-0000-000000000005` version 1 (in realm `f0000000-0000-0000-0000-000000000002`, created by `de10a11c-ec00-1000-0000-000000000000` on 2000-01-07T00:00:00Z): hash digest mismatch!"
```

Reply:
shouldn't it be `hash digestS mismatch` or `hash digest mismatches` ?

Reply:
this code has been removed